waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -507,10 +535,40 @@
           (<= attempt max-patch-retries)
           (recur (inc attempt) (get-replica-count scheduler service-id)))))))
 
-(defn- kill-service-instance
+(let [num-deletion-locks 100
+      deletion-locks (repeatedly num-deletion-locks (fn [] (Object.)))]
+  (defn- service-id->deletion-lock
+    "Returns the deletion lock striped by a hash of the service ID."
+    [service-id]
+    (nth deletion-locks (mod (hash service-id) num-deletion-locks))))
+
+(defn- instance->pod-url
+  "Returns the pod api server url."
+  [api-server-url {:keys [k8s/namespace k8s/pod-name]}]
+  (str api-server-url "/api/v1/namespaces/" namespace "/pods/" pod-name))
+
+(defn hard-delete-service-instance
+  "Force kill the Kubernetes pod corresponding to the given Waiter Service Instance.
+   Does not adjust ReplicaSet replica count; preventing scheduling of a replacement pod must be ensured by the callee.
+   Uses locking to prevent concurrent calls for killing instances from the same service.
+   Returns nil on success, but throws on failure."
+  [{:keys [api-server-url] :as scheduler} {:keys [service-id] :as instance}]
+  (locking
```

Comment:
What problem does this lock solve? The comment says it's to avoid "concurrent calls to kill an instance" — but it's not at all clear to me why it's not OK to do this concurrently while it is OK to do it twice immediately back-to-back (serialized by the lock).

I don't feel like this lock is actually accomplishing anything. Unless we have a concrete scenario where it solves a problem, then I'd prefer to just drop the locks from this change.